### ISSUE TITLE
can't install or run cytoFlow GUI v1.0 on any system

### ISSUE THREAD

**Issue by T5j32pdJ9Mz4** [cytoflow/cytoflow#296]
Hi, thank you for your work. I have a major issue with the gui, as I can't run it.

I tried on several machines, including:

Ubuntu 20.20
cytoflow binary for x86_64 linux v1.0

Crashes at start with error message:

```
An exception has occurred.  Please report a problem from the Help menu!

Afterwards, may need to restart Cytoflow to continue working.

Qt message: Failed to create OpenGL context for format QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize 24, redBufferSize -1, greenBufferSize -1, blueBufferSize -1, alphaBufferSize -1, stencilBufferSize 8, samples 0, swapBehavior QSurfaceFormat::SwapBehavior(DefaultSwapBehavior), swapInterval 1, profile  QSurfaceFormat::OpenGLContextProfile(NoProfile)) 
```

```sh
fc-list: /home/local/tmp/cytoflow-1.0.linux_x86-64/cytoflow-1.0/libuuid.so.1: no version information available (required by /lib/x86_64-linux-gnu/libfontconfig.so.1)
libGL error: MESA-LOADER: failed to open iris (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
libGL error: failed to load driver: iris
libGL error: MESA-LOADER: failed to open iris (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
libGL error: failed to load driver: iris
libGL error: MESA-LOADER: failed to open swrast (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
libGL error: failed to load driver: swrast
fc-list: /home/local/tmp/cytoflow-1.0.linux_x86-64/cytoflow-1.0/libuuid.so.1: no version information available (required by /lib/x86_64-linux-gnu/libfontconfig.so.1)
libGL error: MESA-LOADER: failed to open iris (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
libGL error: failed to load driver: iris
libGL error: MESA-LOADER: failed to open iris (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
libGL error: failed to load driver: iris
libGL error: MESA-LOADER: failed to open swrast (search paths /usr/lib/x86_64-linux-gnu/dri:\$${ORIGIN}/dri:/usr/lib/dri)
libGL error: failed to load driver: swrast
/home/local/tmp/cytoflow-1.0.linux_x86-64/cytoflow-1.0/PyQt5/Qt/libexec/QtWebEngineProcess: /home/local/tmp/cytoflow-1.0.linux_x86-64/cytoflow-1.0/PyQt5/Qt/libexec/../lib/./libuuid.so.1: no version information available (required by /lib/x86_64-linux-gnu/libfontconfig.so.1)
2021-01-26 14:46:22,044 CRITICAL:root:Qt message: Failed to create OpenGL context for format QSurfaceFormat(version 2.0, options QFlags<QSurfaceFormat::FormatOption>(), depthBufferSize 24, redBufferSize -1, greenBufferSize -1, blueBufferSize -1, alphaBufferSize -1, stencilBufferSize 8, samples 0, swapBehavior QSurfaceFormat::SwapBehavior(DefaultSwapBehavior), swapInterval 1, profile  QSurfaceFormat::OpenGLContextProfile(NoProfile)) 
composeAndFlush: makeCurrent() failed
composeAndFlush: makeCurrent() failed
composeAndFlush: makeCurrent() failed
Fatal Python error: Aborted

Thread 0x00007ff8d1ffb640 (most recent call first):
  File "threading.py", line 295 in wait
  File "threading.py", line 551 in wait
  File "cytoflowgui/matplotlib_backend.py", line 184 in send_to_remote
  File "threading.py", line 864 in run
  File "threading.py", line 916 in _bootstrap_inner
  File "threading.py", line 884 in _bootstrap

Thread 0x00007ff8d27fc640 (most recent call first):
  File "selectors.py", line 376 in select
  File "multiprocessing/connection.py", line 911 in wait
  File "multiprocessing/connection.py", line 414 in _poll
  File "multiprocessing/connection.py", line 257 in poll
  File "cytoflowgui/matplotlib_backend.py", line 152 in listen_for_remote
  File "threading.py", line 864 in run
  File "threading.py", line 916 in _bootstrap_inner
  File "threading.py", line 884 in _bootstrap

Thread 0x00007ff8d313d640 (most recent call first):
  File "multiprocessing/connection.py", line 379 in _recv
  File "multiprocessing/connection.py", line 407 in _recv_bytes
  File "multiprocessing/connection.py", line 216 in recv_bytes
  File "multiprocessing/queues.py", line 94 in get
  File "cytoflowgui/workflow.py", line 341 in log_main
  File "threading.py", line 864 in run
  File "threading.py", line 916 in _bootstrap_inner
  File "threading.py", line 884 in _bootstrap

Thread 0x00007ff8d393e640 (most recent call first):
  File "threading.py", line 295 in wait
  File "queue.py", line 164 in get
  File "cytoflowgui/workflow.py", line 331 in send_main
  File "threading.py", line 864 in run
  File "threading.py", line 916 in _bootstrap_inner
  File "threading.py", line 884 in _bootstrap

Thread 0x00007ff901481640 (most recent call first):
  File "selectors.py", line 376 in select
  File "multiprocessing/connection.py", line 911 in wait
  File "multiprocessing/connection.py", line 414 in _poll
  File "multiprocessing/connection.py", line 257 in poll
  File "cytoflowgui/workflow.py", line 272 in recv_main
  File "threading.py", line 864 in run
  File "threading.py", line 916 in _bootstrap_inner
  File "threading.py", line 884 in _bootstrap

Thread 0x00007ff8f5c7e640 (most recent call first):
  File "multiprocessing/popen_fork.py", line 28 in poll
  File "multiprocessing/popen_fork.py", line 50 in wait
  File "multiprocessing/process.py", line 124 in join
  File "cytoflowgui/run.py", line 299 in monitor_remote_process
  File "threading.py", line 864 in run
  File "threading.py", line 916 in _bootstrap_inner
  File "threading.py", line 884 in _bootstrap

Current thread 0x00007ff90a175740 (most recent call first):
  File "site-packages/pyface/ui/qt4/tasks/main_window_layout.py", line 186 in set_layout_for_area
  File "site-packages/pyface/ui/qt4/tasks/main_window_layout.py", line 126 in set_layout
  File "site-packages/pyface/ui/qt4/tasks/task_window_backend.py", line 164 in set_layout
  File "site-packages/pyface/ui/qt4/tasks/task_window_backend.py", line 118 in _layout_state
  File "site-packages/pyface/ui/qt4/tasks/task_window_backend.py", line 71 in show_task
  File "site-packages/pyface/tasks/task_window.py", line 151 in activate_task
  File "site-packages/pyface/tasks/task_window.py", line 332 in set_window_layout
  File "site-packages/envisage/ui/tasks/tasks_application.py", line 228 in create_window
  File "site-packages/envisage/ui/tasks/tasks_application.py", line 305 in _create_windows
  File "site-packages/envisage/ui/tasks/tasks_application.py", line 140 in run
  File "cytoflowgui/cytoflow_application.py", line 124 in run
  File "cytoflowgui/run.py", line 241 in run_gui
  File "cytoflowgui/run.py", line 306 in <module>
Abandon (core dumped)
```
libgl1-mesa-dri and libgl1-mesa-glx are installed (Ubuntu 20.04).
On the linux box, /usr/lib/x86_64-linux-gnu/dri exists, and the file iris_dri.so is in it.
I tried to install mesa-utils, and I can confirm I do have working 3D accelerati
on (spinning gears)

I tried to move libuuid.so.1 out from cytoflow directory, doesn't change the result.

I tried on a fresh CentOS box, same issue.

Please note The same issue is obsevred on a Win10 machine, the software crashes at start, and report missing or broken libraries (Qt-OpenGL relatedi warnings and errors). I also tested recently on a MacOS box, same problem.

**Comment by kristjaneerik** [comment]
You could maybe try the conda-based installation instructions here https://cytoflow.readthedocs.io/en/stable/INSTALL.html

**Comment by T5j32pdJ9Mz4** [comment]
Yes, thank you very much for suggesting this alternative. However, it will not resolve the issue here, which is related to the binaries bundles that are not working.

**Comment by bpteague** [comment]
Hi @T5j32pdJ9Mz4 - unfortunately, this is not a Cytoflow issue but rather an issue with OpenGL on your system. (You are not the first person to report it!) All the references to `libGL` and the MESA loader point to an underlying issue with OpenGL. If you search the web for those error messages, a lot of people receive them using a lot of software -- but I have yet to see anyone with a confirmed fix.

(And you might ask, "OpenGL? Cytoflow doesn't do any 3D graphics." You're right -- but we DO use Qt's web page renderer for showing the help pages, and that depends on OpenGL.)

Finally, @kristjaneerik was not wrong in his suggestion. If you install Anaconda and the `conda` package, it will come with the GUI as well. If this works, but the distributed binary doesn't, I would be VERY interested. If so, please re-open this bug -- but until then, I'm going to close it.

**Comment by bpteague** [comment]
PS - I am going to spin a new release in the next few days. It has updated all the dependencies. Try it too?

**Comment by bpteague** [comment]
I'm .... not sure what to tell you. The error does indeed seem to be an issue with OpenGL. We don't distribute any OpenGL libraries, depending entirely on the local system's libraries. You're not the first person to report this, but you are definitely the first person who has done so with the level of detail you have. Which is why I'm not going to close it off-hand with a "your system's configuration is borked.)

At the same time, I have an entire class of students who just installed it on Windows 10 laptops (with integrated Intel graphics) (and a few Macs) and they all work fine.

Finally, @kristjaneerik wasn't wrong -- if you install Anaconda and the `conda` package, the GUI will be available too. I would be very interested to learn whether it works in that setting or not.

I'm also spinning a new release in the next couple of days. Other than a few of @kristjaneerik's patches, it's a maintenance release that just updates the dependencies. Keep an eye out and try that one too?

**Comment by T5j32pdJ9Mz4** [comment]
Thank you for your detailed answer. As I understood the conda suggestion was a workaround rather than a real fix of the reported issue. I can confirm that the GUI works when launched from a conda env. I understand you do not maintain OpenGL yourself, but maybe there is an issue in the way the dependencies (something version specific) are stated in the GUI binary bundles ? Conda certainly manage it in a more flexible way.
I did test 3 different machines (incl. linux, MacOS, Win10) + 1 vbox, I confirm the binary crashed each time.
I surely will have a try on the next release.
Thanks again.

**Comment by bpteague** [comment]
I'm glad you can get the GUI working from the conda install. That narrows it down to a packaging issue.

I'm wondering if it's actually the very first error that's causing all the rest of them. Could you please just move `libuuid.so.1` to a different directory and then give it a shot?

(I'm trying to reproduce on my own Ubuntu 20.04 system but so far no luck. I'd really like to do so without pulling my graphics card out, if possible!)

**Comment by bpteague** [comment]
(sorry, move the copy of `libuuid.so.1` that's distributed with the GUI bundle, is what I meant.)

**Comment by bpteague** [comment]
Actually ..... there are a BUNCH of OpenGL-related things in that directory. Others to try moving out (so as to make it use the system versions):
- `libEGL.so.1`
- `libglapi.so.0`
- `libxcb-glx.so.0`

:/ man now I'm feeling bad about the other person whose OpenGL configuration I blamed. Again, I really appreciate your being thorough in your bug report. I'd love to get this squashed before the next release. I'll do some more investigating on my end as well.

**Comment by T5j32pdJ9Mz4** [comment]
I tried to move the 4 files out of the cytoflow directory, no more luck on a Ubuntu or CentOS box.

**Comment by bpteague** [comment]
Okay. I'll keep reading/thinking about the Linux situation.

In the meantime, does the Windows error say something about being unable to load libEGL? I ran into an issue with a student today whose laptop wouldn't run the binary distribution. To fix it, we needed to put the following four files in BOTH the main directory (next to `cytoflow.exe`) as well as `PyQt5/Qt/bin`:
- `libEGL.dll`
- `libEGLSv2.dll`
- `d3dcompiler_XX.dll`  (where `XX` is a version number, like `42`)
- `opengl32sw.dll`

(Actually, we probably didn't need to put them in both places -- but we were so relieved to get it working that we didn't stop to figure out which place they actually needed to be.)

Most systems already have these in accessible places due to other software being installed (basically anything that embeds Chrome or Atom), but this student didn't. If this is infact the error you're getting, can you search your system for versions of these files? If you don't find them, let me know and I'll upload them somewhere for you.

**Comment by bpteague** [comment]
I've got the Linux part of this reproducing by installing a fresh Ubuntu 20.04.02 image on a VM. Progress! However, a fresh MacOS (Catalina) install doesn't give me any trouble. Can you give me more detail about the precise issue you're seeing there?

**Comment by bpteague** [comment]
I think I've got this fixed on Linux and Windows:
- On Linux, building with a modern version of Mesa looks in the right place for the system DRI drivers
- On Windows, distributing the ANGLE dependencies as above

On Mac, I can't reproduce. I have one more bug to fix before releasing v1.1 -- if you test 1.1 and it still doesn't work, please re-open.

**Comment by T5j32pdJ9Mz4** [comment]
Hi @bpteague, just a short feedback on the last version. I can confirm that the last binary relase (v1.1) is working on Linux (Ubuntu and CentOS boxes), Win10 and somehow on Mac. I have been able to launch cytoflow v1.1 on Mac Darwin 16.0, but for some reason on Darwin 20.0 the binary package starts but seems to hand on (no graphical output). It could be related to BigSur update, but I am not sure. If I can collect more info about the issue I'll keep you informed.
Again, many thanks for your work on cytoflow and your time investigating the error.
